cip/1.accepted/CIP2017-04-20-query-combinators.adoc

```diff
@@ -0,0 +1,150 @@
+
+`EXCEPT ALL` computes the logical multiset difference between two bags of input records (i.e. excess duplicates on the left-hand side are retained).
+
+=== EXCLUSIVE UNION
+
+`EXCLUSIVE UNION` computes the exclusive logical set union between two sets of input records (i.e. any duplicates in the final outcome are discarded).
+
+`EXCLUSIVE UNION MAX` computes the exclusive logical multiset union between two bags of input records (i.e. the largest remaining excess multiplicity of each record in any argument bag is returned).
+
+=== OTHERWISE
+
+`OTHERWISE` computes the logical choice between two bags of input records.
+It evaluates to all records from the left-hand side argument provided the bag of input records is non-empty; otherwise it evaluates to all records from the right-hand side argument.
+
+=== CROSS
+
+`CROSS` computes the cartesian product between two bags of input records (i.e. preserves duplicates).
+
+In contrast to the other query combinators, the standard rules regarding returned record fields do not apply to `CROSS`.
+Instead, the set of returned record fields of both arms of a `CROSS` must be non-overlapping.
+The returned record fields of a `CROSS` operation consist of all the fields specified in the left arm (appearing in the order specified), followed by all the fields specified in the right arm (appearing in the order specified).
+
+=== THEN
+
+`THEN` computes query-level pipelining; i.e. it executes the right-hand side query for each input record from the left-hand side, and returns the flattened concatenation of all such records produced.
+
+The main feature of `THEN` is that it allows pipelining between nested subqueries.
+This is due to its syntactic status as a query combinator.
+
+In contrast to the other query combinators, the standard rules regarding returned record fields do not apply to `THEN`.
+Instead, the set of returned record fields of both arms of `THEN` may overlap arbitrarily.
+All record fields that are returned in the left arm are made visible at the start of the right-arm query.
+`THEN` returns the record fields that are specified in the right arm, in the order specified in the right arm.
+
+=== Handling of NULL values
+
+All query combinators perform record-level comparisons under equivalence (i.e. `null` is equivalent to `null`).
+
+=== Interaction with existing features
+
+This CIP codifies the pre-existing `UNION` and `UNION ALL` constructs.
+
+The suggested changes are expected to integrate well with the parallel CIP for nested subqueries.
+
+This CIP adds `INTERSECT`, `EXCLUSIVE`, and `OTHERWISE` as new keywords.
+
+=== Alternatives
+
+SQL does not provide `UNION MAX` (it has been suggested in the literature though).
+
+`EXCLUSIVE UNION` and `EXCLUSIVE UNION MAX` are not provided by SQL and could be omitted.
+
+`OTHERWISE` is not provided by SQL and could be omitted.
+
+SQL allows `MINUS` as an alias for `EXCEPT`.
```

Comment:
Acknowledging a potential Pythonic bias, I have a strong preference for `MINUS` over `EXCEPT` here. `EXCEPT` draws to mind exceptions and feels too general as a term. Conversely, `MINUS` leads to mathematical difference, which is entirely appropriate.